### PR TITLE
not_exist_ok for dict_set only allows to add a non existing key to a dictionary, or non existing entry to a list. Not a whole nested structure implied by the query

### PR DESCRIPTION
**Pull request by dafnapension**
closes: https://github.com/IBM/unitxt/issues/1400

Can still set a whole nested structure, but as the value being set, not through processing the query by dict_set.

### CODE REVIEW

**Comment by dafnapension** [comment]
I think these are rather ugly patches.. and I did not manage to cover all. 
I suggest to return back home..

File: prepare/cards/mmmu.py [hunk 1]
```diff
@@ -47,6 +47,7 @@
         ),
         preprocess_steps=[
             RenameSplits(mapper={"dev": "train", "validation": "test"}),
+            SetEmptyDictIfDoesNotExist(field="media"),
```
Comment (yoavkatz) [inline prepare/cards/mmmu.py:50]: I think the media field is used by generic code,  so it should be added by the code and not the cards.

File: src/unitxt/operators.py [hunk 2]
```diff
@@ -277,6 +278,20 @@ def process(
         return instance
 
 
+class SetEmptyDictIfDoesNotExist(InstanceOperator):
```
Comment (yoavkatz) [inline src/unitxt/operators.py:281]: The behavior of this is method is more than expected by the name.  It not only creates an empty dict if it does not exist, but also transform the field from string to dict if string is a json.  I think we should avoid such complex behavior, which is more indicative of a bug somewhere else. 

For example, on Friday I found a bug in post processing pipeline:

image

(Still not in main)

Comment (dafnapension) [reply src/unitxt/operators.py:281]: Hi @yoavkatz , If this helps, I found what triggered my need to add this loads to `SetEmptyDictIfDoesNotExist` (and again, just for patching with גרזנים ולומיק the need to prepare container elements, to not surprise the user):

This method, shamelessly makes `task_data` into a string:
```python
    def serialize_instance_fields(self, instance, task_data):
        if settings.task_data_as_text:
            instance["task_data"] = json.dumps(task_data)

        if not isinstance(instance["source"], str):
            instance["source"] = json.dumps(instance["source"])
        return instance
```

**Comment by dafnapension** [comment]
Hi @yoavkatz , I suspect that the new `SetEmptyDictIfDoesNotExist` that I added, and the about to be added `SetEmptyListIfDoesNotExist` and the enforcement to use them before each and every existing `Set` and more over: each and every existing Operator that writes anything to any `to_field`, will drive our users crazy. Let alone the runtime it will consume.
I suggest to consider any of the following solutions:
(1) extend the possible values of `not_exist_ok` to : No, at_most_last_component, any_missing_component, 
and throw exception accordingly  (my recommendation)
(2) only allow at_most_last_component (that you feel is the decent, expected behavior), and then once an operator and an input instance like those that started the current saga will meet:
applying operator=Copy(field="source", to_field="task_data/source") on input instance
{"source": "hello", "task_data" :3} will throw a Warning, saying something like:
operator Copy employed dict_set in order to set value "hello" to an element (subfield) "task_data/source" that does not exist in the input instance, nor its implied containing subfield of type dict "task_data". (which exists, but is not a dictionary). dict_set is not allowed to generate into the instance elements implied by the query, other than the element implied by the last component. 
dict_set was invoked instead with the following query and value:  dict_set("task_data", {"source" : "hello"}) which sets a value, of type dict, to an existing element, "task_data", replacing its old value.

Other than these explanation and apologies, the computed results will remain the same, users will not have to change their existing code, and/or write tedious future invocations of dict_set s.

**Comment by yoavkatz** [comment]
> Hi @yoavkatz , I suspect that the new `SetEmptyDictIfDoesNotExist` that I added, and the about to be added `SetEmptyListIfDoesNotExist` and the enforcement to use them before each and every existing `Set` and more over: each and every existing Operator that writes anything to any `to_field`, will drive our users crazy. Let alone the runtime it will consume. I suggest to consider any of the following solutions: (1) extend the possible values of `not_exist_ok` to : No, at_most_last_component, any_missing_component, and throw exception accordingly (my recommendation) (2) only allow at_most_last_component (that you feel is the decent, expected behavior), and then once an operator and an input instance like those that started the current saga will meet: applying operator=Copy(field="source", to_field="task_data/source") on input instance {"source": "hello", "task_data" :3} will throw a Warning, saying something like: operator Copy employed dict_set in order to set value "hello" to an element (subfield) "task_data/source" that does not exist in the input instance, nor its implied containing subfield of type dict "task_data". (which exists, but is not a dictionary). dict_set is not allowed to generate into the instance elements implied by the query, other than the element implied by the last component. dict_set was invoked instead with the following query and value: dict_set("task_data", {"source" : "hello"}) which sets a value, of type dict, to an existing element, "task_data", replacing its old value. This seems to achieve the requested result, and still comply with the rule that dict_set changes at most the element identified by the last component of the query.
> 
> Other than these explanation and apologies, the computed results will remain the same, users will not have to change their existing code, and/or write tedious future invocations of dict_set s.

Hi. I guess what I'm missing is how prevalent is this problem and wherher its actually a problem caused by another issue.. I means media and task data are generic fields that  should always be there and added by generic cide. If not, we need to understand why.  

In general, we need to strongly avoid complex flags, that increase the learning curve of the library.

**Comment by dafnapension** [comment]
I agree (although I like flags..).   Can [this issue](https://github.com/IBM/unitxt/pull/1402#discussion_r1865602637) hint to what you are looking for? I bumped into it while needing to push this LoadJson into the patch.

**Comment by dafnapension** [comment]
Too much existing code trust dict_set to fill in all missing component, and not just the last component

**Pull request closed without merging**